packages/promotion/src/utils/compute-actions/shipping-methods.ts

```diff
@@ -0,0 +1,122 @@
+import { PromotionTypes } from "@medusajs/types"
+import {
+  ApplicationMethodAllocation,
+  ApplicationMethodTargetType,
+  MedusaError,
+} from "@medusajs/utils"
+import { areRulesValidForContext } from "../validations"
+
+export function getComputedActionsForShippingMethods(
+  promotion: PromotionTypes.PromotionDTO,
+  shippingMethodApplicationContext: PromotionTypes.ComputeActionContext[ApplicationMethodTargetType.SHIPPING_METHODS],
+  itemIdPromoValueMap: Map<string, number>
+): PromotionTypes.ComputeActions[] {
+  const applicableShippingItems: PromotionTypes.ComputeActionContext[ApplicationMethodTargetType.SHIPPING_METHODS] =
+    []
+
+  if (!shippingMethodApplicationContext) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `"items" not found in context`
+    )
+  }
+
+  for (const shippingMethodContext of shippingMethodApplicationContext) {
+    const isPromotionApplicableToItem = areRulesValidForContext(
+      promotion.application_method?.target_rules!,
+      shippingMethodContext
+    )
+
+    if (!isPromotionApplicableToItem) {
+      continue
+    }
+
+    applicableShippingItems.push(shippingMethodContext)
+  }
+
+  return applyPromotionToShippingMethods(
+    promotion,
+    applicableShippingItems,
+    itemIdPromoValueMap
+  )
+}
+
+export function applyPromotionToShippingMethods(
+  promotion: PromotionTypes.PromotionDTO,
+  shippingMethods: PromotionTypes.ComputeActionContext[ApplicationMethodTargetType.SHIPPING_METHODS],
+  itemIdPromoValueMap: Map<string, number>
+): PromotionTypes.ComputeActions[] {
+  const { application_method: applicationMethod } = promotion
+  const allocation = applicationMethod?.allocation!
+  const computedActions: PromotionTypes.ComputeActions[] = []
+
+  if (allocation === ApplicationMethodAllocation.EACH) {
+    for (const item of shippingMethods!) {
+      const appliedPromoValue = itemIdPromoValueMap.get(item.id) || 0
+      const promotionValue = parseFloat(applicationMethod!.value!)
+      let amount = promotionValue
+      const applicableTotal = item.unit_price - appliedPromoValue
+
+      if (promotionValue > applicableTotal) {
+        amount = applicableTotal
+      }
+
+      if (amount <= 0) {
+        continue
+      }
+
+      itemIdPromoValueMap.set(item.id, appliedPromoValue + amount)
+
+      computedActions.push({
+        action: "addShippingMethodAdjustment",
+        shipping_method_id: item.id,
+        amount,
+        code: promotion.code!,
+      })
+    }
```

Comment:
thought: Refactor to util and use in both `applyPromotionTo...` functions

Reply:
I like this, but would like to keep this until buyget is added. My assumption is that they might differentiate at some point and don't want to do early abstractions here